src/bindings/js.ts

```diff
@@ -789,7 +789,11 @@ export class JSBuilder extends ExportsWalker {
     const
       length = values.length,
       buffer = exports.__pin(exports.__new(length << align, id)) >>> 0;
-    for (let i = 0; i < length; i++) lowerElement(buffer + (i << align >>> 0), values[i]);
+    if (ArrayBuffer.isView(values)) {
+      new values.constructor(memory.buffer, buffer, length).set(values);
```

Comment:
I wonder if this will in all cases work properly. For instance, if `lowerElement` performs some kind of coercion, it would be skipped here, which seems as if it could affect the outcome in different ways depending on whether a typed array or array is provided?

Reply:
Compulsion is needed for untyped data of `Array`. Typed arrays automatically apply coercion during assignment, don't they?

Reply:
I think they would, but since the check here isn't static, but dynamic depending on the kind of value a user provides (can't be sure that a user provides the expected kind of value), isn't this risky?

Reply:
I don't really know what you mean. Can I give you an example? If we `values` is for example `Uint8Array` on host side then the `lowerElement` will look like `(x) => x & 255` and can't contain compound elements like object, nested arrays etc which required chained lowerElement process

Reply:
What I mean that the check here is checking the source type, `values`, which can be anything provided by a user, regardless of what the lowered-to element type is. Say `values` is given as an `Array`, not a typed array, the behavior would be different. Or am I missing something?

Reply:
```
if (ArrayBuffer.isView(values)) { // any Typed Array
  ... fast path ....
} else {  // Array or any other Array Like object
  ... old path ....
}
```

Reply:
Right, so now if the normal path coerces, but the fast path does not, we might get different behavior. For instance, if the static array is an array of managed references, one could write arbitrary numbers into it by triggering the fast path with an incorrect typed array.

Reply:
Typed Array can't contain managed references. Or I miss something?

Reply:
What I mean is that one could just provide one as `values`, even though that's wrong, and no check would be performed.

Reply:
As an example, say the static array is of type `StaticArray<string>`, which doesn't allow `null`s, then one could circumvent the `null` check by incorrectly providing a typed array and writing `null`s.

Reply:
I know what you mean, but users will not use Uint32Array for strings, they will use Array<string> for that. Wouldn't they? That's why everything will remain the same.

Reply:
But they could specify something else, especially when using JS instead of TS, and then there's some sort of undefined behavior in it.

Reply:
For example:
on AS:
```ts
export function foo(arr: StaticArray<string>) { ... }
```
on JS:
```js
exports.foo(["a", "b", "c"]) // will use old path with lowerElement
```
in theory ppl can do this:
```js
exports.foo(Uint32Array.of(1, 2, -4));
```
But this will be wrong in any case. With fast path or without fast path

Reply:
The difference is in

```ts
exports.foo([0, 0, 0]); // or [null, null, null] for that matter
```

and

```ts
exports.foo(new Uint8Array([0, 0, 0]))
```

where the former produces a "thing cannot be null" kind of error and the latter does not, writing zeroes without checking. Am I too careful here? Somehow that looks like a red flag for me.

Reply:
In addition, dts will generate `ArrayLike<string>` which doesn't allow write something like `[0, 0, 0]` or `new Uint8Array([0, 0, 0])`

Reply:
I understand the argument, still something in me thinks that this is one shortcut too many due to introducing magic behavior differences.